larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewLongTracksTool.cc

```diff
@@ -0,0 +1,173 @@
+/**
+ *  @file   larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewLongTracksTool.cc
+ *
+ *  @brief  Implementation of the long tracks tool class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+
+#include "larpandoracontent/LArThreeDReco/LArTwoViewMatching/TwoViewLongTracksTool.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+TwoViewLongTracksTool::TwoViewLongTracksTool() :
+    m_minMatchedFraction(0.3f),
+    m_minMatchingScore(0.95f),
+    m_minMatchedSamplingPoints(15),
+    m_minXOverlapFraction(0.9f),
+    m_minMatchedSamplingPointRatio(2)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewLongTracksTool::HasLongDirectConnections(IteratorList::const_iterator iIter, const IteratorList &iteratorList)
+{
+    for (IteratorList::const_iterator iIter2 = iteratorList.begin(), iIter2End = iteratorList.end(); iIter2 != iIter2End; ++iIter2)
+    {
+        if (iIter == iIter2)
+            continue;
+
+        if (((*iIter)->GetCluster1() == (*iIter2)->GetCluster1()) || ((*iIter)->GetCluster2() == (*iIter2)->GetCluster2()))
+            return true;
+    }
+
+    return false;
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+bool TwoViewLongTracksTool::IsLongerThanDirectConnections(IteratorList::const_iterator iIter, const MatrixType::ElementList &elementList,
+    const unsigned int minMatchedSamplingPointRatio, const pandora::ClusterSet &usedClusters)
+{
+    const unsigned int nMatchedReUpsampledSamplingPoints((*iIter)->GetOverlapResult().GetNMatchedReUpsampledSamplingPoints());
+
+    for (MatrixType::ElementList::const_iterator eIter = elementList.begin(); eIter != elementList.end(); ++eIter)
+    {
+        if ((*iIter) == eIter)
+            continue;
+
+        if (usedClusters.count(eIter->GetCluster1()) || usedClusters.count(eIter->GetCluster2()))
+            continue;
+
+        if (((*iIter)->GetCluster1() != eIter->GetCluster1()) && ((*iIter)->GetCluster2() != eIter->GetCluster2()))
+            continue;
+
+        if (nMatchedReUpsampledSamplingPoints < minMatchedSamplingPointRatio * eIter->GetOverlapResult().GetNMatchedReUpsampledSamplingPoints())
```

Comment:
See later comment about changing `minMatchedSamplingPointRatio` to a float.
This line would become a little ugly to look at:
`static_cast<int>(minMatchedSamplingPointRatio * static_cast<float>(eIter->GetOverlapResult().GetNMatchedReUpsampledSamplingPoints())))`.

There might also need to be some additional protection for `minMatchedSamplingPointRatio` accidentally going negative somewhere...

Reply:
This will be addressed in a future PR!